src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_debuginfod_patch.diff

```diff
@@ -0,0 +1,331 @@
++static debuginfod_end_t the_debuginfod_end = NULL;
++static int debuginfod_guard = 0;
++static int debuginfod_dlopen_failed = 0;
++ 
++static int
++elf_open_debugfile_by_debuginfod (const char *buildid_data,
++			       size_t buildid_size,
++			       backtrace_error_callback error_callback,
++			       void *data)
++{
++  int ret;
++  int does_not_exist;
++
++  if (debuginfod_dlopen_failed) {
++      return -1;
++  }
++
++  if (the_debuginfod_find_debuginfo == NULL) {
++    void *handle = dlopen(DEBUGINFOD_SONAME, RTLD_LAZY);
++    if (!handle) {
++      debuginfod_dlopen_failed = 1;
++      return -1;
++    }
++
++    the_debuginfod_find_debuginfo = (debuginfod_find_debuginfo_t)dlsym(handle, "debuginfod_find_debuginfo");
++    if (!the_debuginfod_find_debuginfo) {
++      return -1;
++    }
++
++    the_debuginfod_begin = (debuginfod_begin_t)dlsym(handle, "debuginfod_begin");
++    if (!the_debuginfod_begin) {
++      return -1;
++    }
++
++    the_debuginfod_end = (debuginfod_end_t)dlsym(handle, "debuginfod_end");
++    if (!the_debuginfod_end) {
++      return -1;
++    }
++  }
++
++  char* path = NULL;
++  const unsigned char* bi = (const unsigned char*)buildid_data;
++
++  debuginfod_client* client = the_debuginfod_begin();
++  if (!client) {
++    return -1;
++  }
++
++  int fd = the_debuginfod_find_debuginfo(client, bi, buildid_size, &path);
++
++  the_debuginfod_end(client);
++
++  if (!path || fd <= 0) {
++    return -1;
++  } 
++
++  ret = backtrace_open (path, error_callback, data, &does_not_exist);
++
++  free(path);
++
++  return ret;
++}
++
++
+ /* Open a separate debug info file, using the build ID to find it.
+    Returns an open file descriptor, or -1.
+ 
+    The GDB manual says that the only place gdb looks for a debug file
+    when the build ID is known is in /usr/lib/debug/.build-id.  */
+-
+ static int
+ elf_open_debugfile_by_buildid (struct backtrace_state *state,
+ 			       const char *buildid_data, size_t buildid_size,
+@@ -4422,10 +4500,17 @@ elf_add (struct backtrace_state *state, const char *filename, int descriptor,
+ 
+       d = elf_open_debugfile_by_buildid (state, buildid_data, buildid_size,
+ 					 error_callback, data);
++      if (d < 0 && !debuginfod_guard) {
++          char* env = getenv(DEBUGINFOD_PROGRESS_ENV_VAR);
++          if (env) {
++            printf("Trying to dowload debuginfo for %s\n", filename);
++          }
++          d = elf_open_debugfile_by_debuginfod(buildid_data, buildid_size,
++                                               error_callback, data);
++      }
+       if (d >= 0)
+ 	{
+ 	  int ret;
+-
+ 	  elf_release_view (state, &buildid_view, error_callback, data);
+ 	  if (debuglink_view_valid)
+ 	    elf_release_view (state, &debuglink_view, error_callback, data);
+@@ -4889,7 +4974,10 @@ backtrace_initialize (struct backtrace_state *state, const char *filename,
+   pd.exe_filename = filename;
+   pd.exe_descriptor = ret < 0 ? descriptor : -1;
+ 
++   
++  debuginfod_guard++;
```

Comment:
Done

Reply:
No, I'm happy enough with this note:

> Note this ONLY affects the symbol resolution when retrieving a backtrace,
> and it doesn't affect offline symbolication.

Since that suggests the solution we'll take if we start getting reports that users find this confusing.